# Taller Electiva VI Pandas 
### Nombre: Alexander Cedeño Rincón
### Asignatura: Electiva VI
### Institución: Univerisdad Antonio Nariño/Sede Neiva

1. importe pandas usando el alias pd.

In [1]:
import pandas as pd

2. imprima la version de pandas que ha sido importada.

In [2]:
print(pd.__version__)
pd.show_versions()

1.0.5


C:\ProgramData\Anaconda3\lib\site-packages\setuptools\distutils_patch.py:25: UserWarning: Distutils was imported before Setuptools. This usage is discouraged and may exhibit undesirable behaviors or errors. Please use Setuptools' objects directly or at least import Setuptools first.
  warnings.warn(



INSTALLED VERSIONS
------------------
commit           : None
python           : 3.8.3.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
machine          : AMD64
processor        : Intel64 Family 6 Model 126 Stepping 5, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : Spanish_Colombia.1252

pandas           : 1.0.5
numpy            : 1.18.5
pytz             : 2020.1
dateutil         : 2.8.1
pip              : 20.1.1
setuptools       : 49.2.0.post20200714
Cython           : 0.29.21
pytest           : 5.4.3
hypothesis       : None
sphinx           : 3.1.2
blosc            : None
feather          : None
xlsxwriter       : 1.2.9
lxml.etree       : 4.5.2
html5lib         : 1.1
pymysql          : None
psycopg2         : None
jinja2           : 2.11.2
IPython          : 7.16.1
pandas_datareader: None
bs4              : 4.9.1
bottleneck       : 1.3.2
fastparquet      : None
gcsfs            : None
lxml.etree

4. cree un dataframe df a partir del diccionario data con los indices de la lista labels.

In [3]:
import numpy as np

data = {'animal': ['gato', 'gato', 'serpiente', 'perro', 'perro', 'gato', 'serpiente', 'gato', 'perro', 'perro'],
        'edad': [2.5, 3, 0.5, np.nan, 5, 2, 4.5, np.nan, 7, 3],
        'visitas': [1, 3, 2, 3, 2, 3, 1, 1, 2, 1],
        'prioridad': ['si', 'si', 'no', 'si', 'no', 'no', 'no', 'si', 'no', 'no']}

labels = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']
print(data)
df= pd.DataFrame({'labels':['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j'],'animal': ['gato', 'gato', 'serpiente', 'perro', 'perro', 'gato', 'serpiente', 'gato', 'perro', 'perro'],
        'edad': [2.5, 3, 0.5, np.nan, 5, 2, 4.5, np.nan, 7, 3],
        'visitas': [1, 3, 2, 3, 2, 3, 1, 1, 2, 1],
        'prioridad': ['si', 'si', 'no', 'si', 'no', 'no', 'no', 'si', 'no', 'no']})
print (df)



{'animal': ['gato', 'gato', 'serpiente', 'perro', 'perro', 'gato', 'serpiente', 'gato', 'perro', 'perro'], 'edad': [2.5, 3, 0.5, nan, 5, 2, 4.5, nan, 7, 3], 'visitas': [1, 3, 2, 3, 2, 3, 1, 1, 2, 1], 'prioridad': ['si', 'si', 'no', 'si', 'no', 'no', 'no', 'si', 'no', 'no']}
  labels     animal  edad  visitas prioridad
0      a       gato   2.5        1        si
1      b       gato   3.0        3        si
2      c  serpiente   0.5        2        no
3      d      perro   NaN        3        si
4      e      perro   5.0        2        no
5      f       gato   2.0        3        no
6      g  serpiente   4.5        1        no
7      h       gato   NaN        1        si
8      i      perro   7.0        2        no
9      j      perro   3.0        1        no


5. Muestre un resumen de la información básica sobre este DataFrame y sus datos (pista: hay un método único al que se puede llamar en el DataFrame).

In [4]:
df.describe()

edad    visitas
count  8.000000  10.000000
mean   3.437500   1.900000
std    2.007797   0.875595
min    0.500000   1.000000
25%    2.375000   1.000000
50%    3.000000   2.000000
75%    4.625000   2.750000
max    7.000000   3.000000

6. visualice las primeras 3 filas del DataFrame df

In [5]:
df.head(3)


labels     animal  edad  visitas prioridad
0      a       gato   2.5        1        si
1      b       gato   3.0        3        si
2      c  serpiente   0.5        2        no

7. Seleccione solo las columnas 'animal' y 'edad' del DataFrame df.

In [6]:
df.iloc[:, 0:2]

labels     animal
0      a       gato
1      b       gato
2      c  serpiente
3      d      perro
4      e      perro
5      f       gato
6      g  serpiente
7      h       gato
8      i      perro
9      j      perro

8. Seleccione los datos en las filas [3, 4, 8] en las columnas ['animal', 'edad'].

In [7]:
df.iloc[[3,4,8], 0:2]

labels animal
3      d  perro
4      e  perro
8      i  perro

9. Seleccione solo las filas en las que el número de visitas sea superior a 3.

No hay filas con visitas superior a 3

10. Seleccione las filas donde falta la edad, es decir, es NaN

In [8]:
df.iloc[[3,7]]

labels animal  edad  visitas prioridad
3      d  perro   NaN        3        si
7      h   gato   NaN        1        si

11. Seleccione las filas donde el animal es un gato y la edad es menor de 3.

In [9]:
df.iloc[[0,5]]

labels animal  edad  visitas prioridad
0      a   gato   2.5        1        si
5      f   gato   2.0        3        no

12. Seleccione las filas cuya edad está entre 2 y 4 (inclusive).

In [10]:
df.iloc[[0,1,5,9]]

labels animal  edad  visitas prioridad
0      a   gato   2.5        1        si
1      b   gato   3.0        3        si
5      f   gato   2.0        3        no
9      j  perro   3.0        1        no

13. Cambie la edad en la fila f a 1.5.

In [11]:
df.replace({'edad': 2}, 1.5)

labels     animal  edad  visitas prioridad
0      a       gato   2.5        1        si
1      b       gato   3.0        3        si
2      c  serpiente   0.5        2        no
3      d      perro   NaN        3        si
4      e      perro   5.0        2        no
5      f       gato   1.5        3        no
6      g  serpiente   4.5        1        no
7      h       gato   NaN        1        si
8      i      perro   7.0        2        no
9      j      perro   3.0        1        no

15. Calcule la edad media de cada animal diferente en df

In [12]:
Edad_Media_Gatos = df[df['animal'] == 'gato']
Edad_Media_Gatos.head()
Edad_Media_Gatos['edad'].describe()

count    3.00
mean     2.50
std      0.50
min      2.00
25%      2.25
50%      2.50
75%      2.75
max      3.00
Name: edad, dtype: float64

In [13]:
Edad_Media_Perros = df[df['animal'] == 'perro']
Edad_Media_Perros.head()
Edad_Media_Perros['edad'].describe()

count    3.0
mean     5.0
std      2.0
min      3.0
25%      4.0
50%      5.0
75%      6.0
max      7.0
Name: edad, dtype: float64

In [14]:
Edad_Media_Serpientes = df[df['animal'] == 'serpiente']
Edad_Media_Serpientes.head()
Edad_Media_Serpientes['edad'].describe()

count    2.000000
mean     2.500000
std      2.828427
min      0.500000
25%      1.500000
50%      2.500000
75%      3.500000
max      4.500000
Name: edad, dtype: float64

16. Agrega una nueva fila 'k' a df con tu elección de valores para cada columna. Luego elimine esa fila para asi volver al DataFrame original

In [15]:
df.loc['k']=['k','gato',1,2,'si']
print(df)
df.drop(['k'],axis=0)
#k= pd.DataFrame({'labels':['k'],
                # 'animal':['gato'],
                # 'edad':[1],
                 #'visitas':[2],
               #  'prioridad':['si']})
#print (k)
#ndf=pd.concat([df,k])
#k.index = range(k.shape[0])
#print(ndf)
#ndf.drop([0],axis=0)

  labels     animal  edad  visitas prioridad
0      a       gato   2.5        1        si
1      b       gato   3.0        3        si
2      c  serpiente   0.5        2        no
3      d      perro   NaN        3        si
4      e      perro   5.0        2        no
5      f       gato   2.0        3        no
6      g  serpiente   4.5        1        no
7      h       gato   NaN        1        si
8      i      perro   7.0        2        no
9      j      perro   3.0        1        no
k      k       gato   1.0        2        si


labels     animal  edad  visitas prioridad
0      a       gato   2.5        1        si
1      b       gato   3.0        3        si
2      c  serpiente   0.5        2        no
3      d      perro   NaN        3        si
4      e      perro   5.0        2        no
5      f       gato   2.0        3        no
6      g  serpiente   4.5        1        no
7      h       gato   NaN        1        si
8      i      perro   7.0        2        no
9      j      perro   3.0        1        no

17. Cuente el número de cada tipo de animal en df

In [16]:
df= pd.DataFrame({'labels':['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j'],'animal': ['gato', 'gato', 'serpiente', 'perro', 'perro', 'gato', 'serpiente', 'gato', 'perro', 'perro'],
        'edad': [2.5, 3, 0.5, np.nan, 5, 2, 4.5, np.nan, 7, 3],
        'visitas': [1, 3, 2, 3, 2, 3, 1, 1, 2, 1],
        'prioridad': ['si', 'si', 'no', 'si', 'no', 'no', 'no', 'si', 'no', 'no']})

Numero_Animales = df['animal'].value_counts()
print(Numero_Animales)



gato         4
perro        4
serpiente    2
Name: animal, dtype: int64


18. Ordene df primero por los valores en la 'edad' en orden descendente, luego por el valor en la columna 'visitas' en orden ascendente (por lo que la fila i debe ser la primera y la fila d debe ser la última).

In [17]:
df_Ordenado = df.sort_values(by= ['edad','visitas'], ascending = [False, True], inplace= False)
print (df_Ordenado)

  labels     animal  edad  visitas prioridad
8      i      perro   7.0        2        no
4      e      perro   5.0        2        no
6      g  serpiente   4.5        1        no
9      j      perro   3.0        1        no
1      b       gato   3.0        3        si
0      a       gato   2.5        1        si
5      f       gato   2.0        3        no
2      c  serpiente   0.5        2        no
7      h       gato   NaN        1        si
3      d      perro   NaN        3        si


19. La columna 'prioridad' contiene los valores 'sí' y 'no'. Reemplace esta columna con una columna de valores booleanos: 'sí' debe ser True y 'no' debe ser False.

In [18]:
df.replace({'prioridad': {'si': 'True', 'no': 'False'}})

labels     animal  edad  visitas prioridad
0      a       gato   2.5        1      True
1      b       gato   3.0        3      True
2      c  serpiente   0.5        2     False
3      d      perro   NaN        3      True
4      e      perro   5.0        2     False
5      f       gato   2.0        3     False
6      g  serpiente   4.5        1     False
7      h       gato   NaN        1      True
8      i      perro   7.0        2     False
9      j      perro   3.0        1     False

20. En la columna 'animal', cambie las entradas de 'serpiente' a 'python'.

In [19]:
df.replace({'animal': 'serpiente'}, 'python')

labels  animal  edad  visitas prioridad
0      a    gato   2.5        1        si
1      b    gato   3.0        3        si
2      c  python   0.5        2        no
3      d   perro   NaN        3        si
4      e   perro   5.0        2        no
5      f    gato   2.0        3        no
6      g  python   4.5        1        no
7      h    gato   NaN        1        si
8      i   perro   7.0        2        no
9      j   perro   3.0        1        no

21. Para cada tipo de animal y cada número de visitas, encuentre la edad media. En otras palabras, cada fila es un animal, cada columna es un número de visitas y los valores son las edades medias

In [20]:
Table= pd.pivot_table(df, values= 'edad', index='animal', columns = 'visitas', aggfunc=np.mean)
Table

visitas      1    2    3
animal                  
gato       2.5  NaN  2.5
perro      3.0  6.0  NaN
serpiente  4.5  0.5  NaN

22. Se tiene un DataFrame df con una columna 'A' de enteros.

df = pd.DataFrame({'A': [1, 2, 2, 3, 4, 5, 5, 5, 6, 7, 7]})
¿Cómo se filtran las filas que contienen el mismo número entero que la fila inmediatamente superior?

Debería quedarse con una columna que contiene los siguientes valores:

1, 2, 3, 4, 5, 6, 7

In [21]:
df1 = pd.DataFrame({'A': [1, 2, 2, 3, 4, 5, 5, 5, 6, 7, 7]})
df1
df1.drop_duplicates(['A'], keep='last')# keep me deja la ultima fila del duplicado.

A
0   1
2   2
3   3
4   4
7   5
8   6
10  7

23. Dado un DataFrame de valores numéricos:

df = pd.DataFrame(np.random.random(size=(5, 3))) # tabla de valores reales de 5x3  
¿Cómo se resta la media de la fila de cada elemento de la fila?

In [22]:
df2 = pd.DataFrame(np.random.random(size=(5, 3))) # tabla de valores reales de 5x3
print (df2)
x= df2.mean(axis=1)
print(x)
df2.sub(df2.mean(axis=1), axis=0)# Restar por filas y promediar por filas 


          0         1         2
0  0.748649  0.844809  0.239596
1  0.974885  0.074693  0.324705
2  0.236743  0.148296  0.534978
3  0.231297  0.557342  0.768859
4  0.512409  0.873147  0.697529
0    0.611018
1    0.458094
2    0.306672
3    0.519166
4    0.694361
dtype: float64


0         1         2
0  0.137631  0.233791 -0.371422
1  0.516791 -0.383401 -0.133389
2 -0.069929 -0.158377  0.228306
3 -0.287869  0.038176  0.249693
4 -0.181953  0.178785  0.003167

24. Suponga que tiene DataFrame con 10 columnas de números reales, por ejemplo:

df = pd.DataFrame(np.random.random(size=(5, 10)), columns=list('abcdefghij'))
¿Qué columna de números tiene la suma más pequeña? Devuelva la etiqueta de esa columna.

In [23]:
df3 = pd.DataFrame(np.random.random(size=(5, 10)), columns=list('abcdefghij')) 
print(df3)
mini= df3.sum()
print(mini)
mini.min()


          a         b         c         d         e         f         g  \
0  0.737535  0.680101  0.399057  0.193373  0.157550  0.270581  0.045425   
1  0.525402  0.557920  0.295590  0.596603  0.018935  0.070324  0.430354   
2  0.123369  0.981940  0.265339  0.960083  0.164181  0.063546  0.060315   
3  0.100904  0.738360  0.885586  0.908650  0.409951  0.704875  0.186293   
4  0.493983  0.119742  0.189073  0.095976  0.806866  0.021299  0.951568   

          h         i         j  
0  0.427303  0.942278  0.651288  
1  0.602955  0.993751  0.877328  
2  0.555747  0.452117  0.666676  
3  0.766346  0.361727  0.709698  
4  0.605947  0.789959  0.100653  
a    1.981192
b    3.078063
c    2.034645
d    2.754685
e    1.557484
f    1.130625
g    1.673956
h    2.958299
i    3.539832
j    3.005642
dtype: float64


1.1306247488429717

25. ¿Cómo se cuenta cuántas filas únicas tiene un DataFrame (es decir, ignora todas las filas que están duplicadas)? Como entrada, use un DataFrame de ceros y unos con 10 filas y 3 columnas.

df = pd.DataFrame(np.random.randint(0, 2, size=(10, 3)))

In [24]:
df4 = pd.DataFrame(np.random.randint(0, 2, size=(10, 3)))
F_U = df4.iloc[:,[0,1,2]]
print(F_U)
New = F_U.drop_duplicates(keep = 'last')
print(New)


   0  1  2
0  0  0  1
1  1  1  1
2  1  0  1
3  1  0  0
4  1  1  1
5  1  0  0
6  0  1  0
7  0  1  1
8  1  1  0
9  1  0  1
   0  1  2
0  0  0  1
4  1  1  1
5  1  0  0
6  0  1  0
7  0  1  1
8  1  1  0
9  1  0  1


26. En la celda a continuación, se tiene un DataFrame df que consta de 10 columnas de números reales. De esos exactamente 5 entradas en cada fila son valores de NaN.

Para cada fila del DataFrame, busque la columna que contiene el tercer valor de NaN.

Debe devolver una columna de etiquetas de la forma: e, c, d, h, d

In [25]:
nan = np.nan

data = [[0.04,  nan,  nan, 0.25,  nan, 0.43, 0.71, 0.51,  nan,  nan],
        [ nan,  nan,  nan, 0.04, 0.76,  nan,  nan, 0.67, 0.76, 0.16],
        [ nan,  nan, 0.5 ,  nan, 0.31, 0.4 ,  nan,  nan, 0.24, 0.01],
        [0.49,  nan,  nan, 0.62, 0.73, 0.26, 0.85,  nan,  nan,  nan],
        [ nan,  nan, 0.41,  nan, 0.05,  nan, 0.61,  nan, 0.48, 0.68]]

columnas = list('abcdefghij')

df5 = pd.DataFrame(data, columns=columnas)
print (df5)

# write a solution to the question here
(df5.isnull().cumsum(axis=1)==3).idxmax(axis=1)

      a   b     c     d     e     f     g     h     i     j
0  0.04 NaN   NaN  0.25   NaN  0.43  0.71  0.51   NaN   NaN
1   NaN NaN   NaN  0.04  0.76   NaN   NaN  0.67  0.76  0.16
2   NaN NaN  0.50   NaN  0.31  0.40   NaN   NaN  0.24  0.01
3  0.49 NaN   NaN  0.62  0.73  0.26  0.85   NaN   NaN   NaN
4   NaN NaN  0.41   NaN  0.05   NaN  0.61   NaN  0.48  0.68


0    e
1    c
2    d
3    h
4    d
dtype: object

27. Implemente una media móvil el grupos con tamaño de ventana 3, que ignora el valor de NaN. Por ejemplo, considere el siguiente DataFrame::

>>> df = pd.DataFrame({'group': list('aabbabbbabab'),
                       'value': [1, 2, 3, np.nan, 2, 3, np.nan, 1, 7, 3, np.nan, 8]})
>>> df
   group  value
0      a    1.0
1      a    2.0
2      b    3.0
3      b    NaN
4      a    2.0
5      b    3.0
6      b    NaN
7      b    1.0
8      a    7.0
9      b    3.0
10     a    NaN
11     b    8.0
The goal is to compute the Series:

0     1.000000
1     1.500000
2     3.000000
3     3.000000
4     1.666667
5     3.000000
6     3.000000
7     2.000000
8     3.666667
9     2.000000
10    4.500000
11    4.000000
P.ej. la primera ventana de tamaño tres para el grupo 'b' tiene valores 3.0, NaN y 3.0 y ocurre en el índice de fila 5. En lugar de ser NaN, el valor en la nueva columna en este índice de fila debe ser 3.0 (solo los dos valores que no son NaN se utilizan para calcular la media (3 + 3) / 2)

In [27]:
df6 = pd.DataFrame({'group': list('aabbabbbabab'),
                       'value': [1, 2, 3, np.nan, 2, 3, np.nan, 1, 7, 3, np.nan, 8]})
print (df6)
dfg=df6.groupby(['group'])['value']
valc= df6.fillna(0).groupby(['group'])['value']
s= valc.rolling(3, min_periods=1).sum()/dfg.rolling(3, min_periods=1).count()
s
s.reset_index(level=0, drop=True).sort_index()


   group  value
0      a    1.0
1      a    2.0
2      b    3.0
3      b    NaN
4      a    2.0
5      b    3.0
6      b    NaN
7      b    1.0
8      a    7.0
9      b    3.0
10     a    NaN
11     b    8.0


0     1.000000
1     1.500000
2     3.000000
3     3.000000
4     1.666667
5     3.000000
6     3.000000
7     2.000000
8     3.666667
9     2.000000
10    4.500000
11    4.000000
Name: value, dtype: float64